engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java

```diff
@@ -214,19 +217,35 @@ private FunctionCallInfo functionCallInfoById(UUID nodeId) {
     return calls.get(nodeId);
   }
 
-  private String[] typeOf(Object value) {
+  private TypeInfo typeOf(Object value) {
     if (value instanceof UnresolvedSymbol) {
-      return new String[] {Constants.UNRESOLVED_SYMBOL};
+      return TypeInfo.ofType(Constants.UNRESOLVED_SYMBOL);
     }
 
-    var typeOfNode = TypeOfNode.getUncached();
-    Type[] allTypes = value == null ? null : typeOfNode.findAllTypesOrNull(value, true);
-    if (allTypes != null) {
-      String[] result = new String[allTypes.length];
-      for (var i = 0; i < allTypes.length; i++) {
-        result[i] = getTypeQualifiedName(allTypes[i]);
+    if (value != null) {
+      final TypeOfNode typeOfNode = TypeOfNode.getUncached();
+      final Type[] publicTypes = typeOfNode.findAllTypesOrNull(value, false);
+
+      if (publicTypes != null) {
+        final Type[] allTypes = typeOfNode.findAllTypesOrNull(value, true);
+        // Relies on the fact that allTypes appends extra types to the end of publicTypes.
+        assert (publicTypes.length == 0
+            || Objects.equals(
+                publicTypes[publicTypes.length - 1], allTypes[publicTypes.length - 1]));
+        final Type[] conversionTypes =
+            Arrays.copyOfRange(allTypes, publicTypes.length, allTypes.length);
+
+        final String[] publicTypeNames = new String[publicTypes.length];
+        for (var i = 0; i < publicTypes.length; i++) {
+          publicTypeNames[i] = getTypeQualifiedName(publicTypes[i]);
+        }
+        final String[] conversionTypeNames = new String[conversionTypes.length];
+        for (var i = 0; i < conversionTypeNames.length; i++) {
+          conversionTypeNames[i] = getTypeQualifiedName(conversionTypes[i]);
+        }
+
+        return new TypeInfo(publicTypeNames, conversionTypeNames);
```

Comment:
```suggestion
        return new TypeInfo(publicTypeNames, hiddenTypeNames);
```
I guess we want to be consistent and `TypeInfo` is using _hidden_ terminology.